### DISCUSSION TITLE
Format of .asset files and .dll logic

### DISCUSSION THREAD

**Discussion by aenrione** [XboxUnity/AuroraAssetEditor#20] [Q&A] [answered]
Hello!
I am trying to build a tool that can convert images and assets back and forth in a generic language like Python or JS in order to build a tool that is cross-platform so I can change Auroras covers through linux or any other frontend API. I am trying to implement the logic you have on `AuroraAsset.cs` and `AuroraAssetDll.cs` but I am not sure I understand what the DLL does under the hood without correctly understanding the hidden format.

Is there a chance you could help better understand the format so I can implement this tool? Any help would be much appreciated!!

**Comment by Swizzy** [comment]
It is converting to/from Direct3D surfaces basically - there is a plan to at somepoint update this with newer .NET versions which are cross platform compatible and remove the dependency on the Windows DLL

**Reply by aenrione** [reply]
Do we know a little bit more about the Direct3D surfaces? Is the Dll using some kine of compression like DXT1-5 or anything like that?

from what i see in the code the format goes something like this:

1. File Header (12 bytes):
- Magic number (4 bytes) - "AEXR" (0x52584541 in reverse)
- Version (4 bytes) - Currently only version 1 is supported
- Data size (4 bytes) - Total size of asset data

2. Entry Table Header (8 bytes):
- Flags (4 bytes) - Bitmap indicating which assets are present
- Screenshot count (4 bytes) - Number of screenshots

3. Entry Table:
- Array of entries, one for each asset type (Icon, Banner, Boxart, etc.)
- Each entry is 64 bytes:
  - Offset (4 bytes) - Offset to asset data
  - Size (4 bytes) - Size of asset data
  - Extended Info (4 bytes) - Currently unused
  - Texture Header (52 bytes) - Direct3D texture information

4. Padding:
- Padding to align data to 2048 byte boundary

5. Asset Data:
- Consecutive blocks of compressed texture data
- Each block corresponds to an entry in the table
- Data appears to be Direct3D compressed textures

The key conversion processes are:

1. Image to Asset:
- Takes raw ARGB pixel data
- Compresses it using Direct3D texture compression
- Generates appropriate texture header
- Outputs compressed data and header

2. Asset to Image:
- Takes compressed texture data and header
- Decompresses Direct3D texture format
- Converts to raw ARGB pixel data

My question is that if this is somewhat correct and if there's anything relevant you could tell me about the Direct3D textures

**Reply by aenrione** [reply]
Also, is the compression really necessary for .asset files? could aurora just read raw ARGB pixel data at the cost of higher file sizes?

**Comment by Swizzy** [comment, marked as answer]
The format of the .asset files is defined/documented here; https://github.com/XboxUnity/AuroraAssetEditor/blob/master/010%20Editor%20Binary%20Template/AuroraAssetTemplate.bt i missed your question about that part.

Reading it again, i noticed there is quite a bit of documentation there on the format of the Direct3D header atleast.

I believe you need to have compression enabled and that it's DXT5 maybe?

**Comment by jrobiche** [comment]
I know that this discussion has been marked as answered, but I wanted to provide some additional information for anyone that may come across this.

Basic documentation on structs in `.asset` files can be found here:
https://github.com/jrobiche/xbox360-aurora-developer-documentation/blob/main/docs/aurora-0.7b.2r1655-assets/README.md
(I need to update some of the information on this page but it should be a solid starting point)

Open source (GPL-v3-or-later) Rust implementation of importing and exporting images to/from `.asset` files can be found here:
https://github.com/jrobiche/libaustralis/blob/7bba3345b86d97a431ab874c77841be3ec528e34/src/aurora/assets.rs#L220
https://github.com/jrobiche/libaustralis/blob/7bba3345b86d97a431ab874c77841be3ec528e34/src/aurora/assets.rs#L174

Feel free to ask if you have any specific questions. I can try to answer them

**Reply by aenrione** [reply]
That is incredibly helpful. Thank you so much!
